File: python/trigger_orders.py
"""
Trigger Orders Example

Stop loss and take profit orders.

Requires: PRIVATE_KEY environment variable
"""

import os
from hyperliquid_sdk import HyperliquidSDK, Side

private_key = os.environ.get("PRIVATE_KEY")
if not private_key:
    print("Set PRIVATE_KEY environment variable")
    print("Example: export PRIVATE_KEY='0x...'")
    exit(1)

sdk = HyperliquidSDK(private_key=private_key)
mid = sdk.get_mid("BTC")
print(f"BTC mid: ${mid:,.2f}")

# Stop loss order (market) - triggers when price falls below stop price
# No limit_price means market order when triggered
# result = sdk.stop_loss(
#     "BTC",
#     size=0.001,
#     trigger_price=mid * 0.95,  # 5% below current
# )
# print(f"Stop loss (market): {result}")

# Stop loss order (limit) - triggers and places limit order at limit_price
# result = sdk.stop_loss(
#     "BTC",
#     size=0.001,
#     trigger_price=mid * 0.95,
#     limit_price=mid * 0.94,
# )
# print(f"Stop loss (limit): {result}")

# Take profit order (market) - triggers when price rises above trigger
# result = sdk.take_profit(
#     "BTC",
#     size=0.001,
#     trigger_price=mid * 1.05,  # 5% above current
# )
# print(f"Take profit (market): {result}")

# Take profit order (limit)
# result = sdk.take_profit(
#     "BTC",
#     size=0.001,
#     trigger_price=mid * 1.05,
#     limit_price=mid * 1.06,
# )
# print(f"Take profit (limit): {result}")

# For buy-side stop/TP (e.g., closing a short position), use side=Side.BUY
# result = sdk.stop_loss("BTC", size=0.001, trigger_price=mid * 1.05, side=Side.BUY)

print("\nTrigger order methods available:")
print("  sdk.stop_loss(asset, size=, trigger_price=, limit_price=None, side=Side.SELL)")
print("  sdk.take_profit(asset, size=, trigger_price=, limit_price=None, side=Side.SELL)")
print("  sdk.trigger_order(TriggerOrder(...))")
print("\nNote: Omit limit_price for market orders when triggered")

File: typescript/trigger_orders.ts
#!/usr/bin/env npx ts-node
/**
 * Trigger Orders Example
 *
 * Stop loss and take profit orders.
 *
 * Requires: PRIVATE_KEY environment variable
 */

import { HyperliquidSDK, Side } from 'hyperliquid-sdk';

async function main() {
  const privateKey = process.env.PRIVATE_KEY;
  if (!privateKey) {
    console.log("Set PRIVATE_KEY environment variable");
    console.log("Example: export PRIVATE_KEY='0x...'");
    process.exit(1);
  }

  const sdk = new HyperliquidSDK(undefined, { privateKey });
  const mid = await sdk.getMid("BTC");
  console.log(`BTC mid: $${mid.toLocaleString()}`);

  // Stop loss order (market) - triggers when price falls below stop price
  // No limitPrice means market order when triggered
  // const result = await sdk.stopLoss("BTC", {
  //   size: 0.001,
  //   triggerPrice: mid * 0.95,  // 5% below current
  // });
  // console.log(`Stop loss (market): ${JSON.stringify(result)}`);

  // Stop loss order (limit) - triggers and places limit order at limitPrice
  // const result = await sdk.stopLoss("BTC", {
  //   size: 0.001,
  //   triggerPrice: mid * 0.95,
  //   limitPrice: mid * 0.94,
  // });
  // console.log(`Stop loss (limit): ${JSON.stringify(result)}`);

  // Take profit order (market) - triggers when price rises above trigger
  // const result = await sdk.takeProfit("BTC", {
  //   size: 0.001,
  //   triggerPrice: mid * 1.05,  // 5% above current
  // });
  // console.log(`Take profit (market): ${JSON.stringify(result)}`);

  // Take profit order (limit)
  // const result = await sdk.takeProfit("BTC", {
  //   size: 0.001,
  //   triggerPrice: mid * 1.05,
  //   limitPrice: mid * 1.06,
  // });
  // console.log(`Take profit (limit): ${JSON.stringify(result)}`);

  // For buy-side stop/TP (e.g., closing a short position), use side: Side.BUY
  // const result = await sdk.stopLoss("BTC", { size: 0.001, triggerPrice: mid * 1.05, side: Side.BUY });

  console.log("\nTrigger order methods available:");
  console.log("  sdk.stopLoss(asset, { size, triggerPrice, limitPrice?, side? })");
  console.log("  sdk.takeProfit(asset, { size, triggerPrice, limitPrice?, side? })");
  console.log("  sdk.triggerOrder(TriggerOrder(...))");
  console.log("\nNote: Omit limitPrice for market orders when triggered");
}

main().catch(console.error);
